File: java/dbos-starter/app/src/test/java/org/example/AppTest.java
package org.example;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

import dev.dbos.transact.DBOS;
import dev.dbos.transact.execution.ThrowingRunnable;

import org.junit.jupiter.api.Test;
import org.mockito.ArgumentMatchers;
import org.mockito.InOrder;
import org.mockito.Mockito;

class AppTest {

  @Test
  void exampleWorkflowExecutesAllStepsInOrder() throws Exception {
    // Arrange
    var mockDBOS = mock(DBOS.class);
    var service = new DurableStarterServiceImpl(mockDBOS);

    // Act
    service.exampleWorkflow();

    // Assert - Verify all steps are executed in correct order
    InOrder inOrder = Mockito.inOrder(mockDBOS);

    // Verify step execution and event setting sequence
    inOrder
        .verify(mockDBOS)
        .runStep(ArgumentMatchers.<ThrowingRunnable<RuntimeException>>any(), eq("stepOne"));
    inOrder.verify(mockDBOS).setEvent(eq("steps_event"), eq(1));
    inOrder
        .verify(mockDBOS)
        .runStep(ArgumentMatchers.<ThrowingRunnable<RuntimeException>>any(), eq("stepTwo"));
    inOrder.verify(mockDBOS).setEvent(eq("steps_event"), eq(2));
    inOrder
        .verify(mockDBOS)
        .runStep(ArgumentMatchers.<ThrowingRunnable<RuntimeException>>any(), eq("stepThree"));
    inOrder.verify(mockDBOS).setEvent(eq("steps_event"), eq(3));

    // Verify total interaction counts
    verify(mockDBOS, times(3))
        .runStep(ArgumentMatchers.<ThrowingRunnable<RuntimeException>>any(), any(String.class));
    verify(mockDBOS, times(3)).setEvent(eq("steps_event"), any(Integer.class));
  }
}

File: kotlin/dbos-starter/app/src/test/kotlin/org/example/AppTest.kt
package org.example

import dev.dbos.transact.*
import dev.dbos.transact.execution.ThrowingSupplier
import kotlin.test.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyString
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mockito.*

class AppTest {

  @Test
  fun exampleWorkflowExecutesAllStepsInOrder() {
    // Arrange
    val mockDBOS = mock<DBOS>()
    val service = DurableStarterServiceImpl(mockDBOS)

    // Act
    service.exampleWorkflow()

    // Assert - Verify all steps are executed in correct order
    val inOrder = inOrder(mockDBOS)

    // Verify step execution and event setting sequence
    // Note: Using more flexible matchers since Kotlin extensions convert lambdas differently
    inOrder.verify(mockDBOS).runStep(any<ThrowingSupplier<Unit, Exception>>(), eq("stepOne"))
    inOrder.verify(mockDBOS).setEvent(eq(STEPS_EVENT), eq(1))
    inOrder.verify(mockDBOS).runStep(any<ThrowingSupplier<Unit, Exception>>(), eq("stepTwo"))
    inOrder.verify(mockDBOS).setEvent(eq(STEPS_EVENT), eq(2))
    inOrder.verify(mockDBOS).runStep(any<ThrowingSupplier<Unit, Exception>>(), eq("stepThree"))
    inOrder.verify(mockDBOS).setEvent(eq(STEPS_EVENT), eq(3))

    // Verify total interaction counts
    verify(mockDBOS, times(3)).runStep(any<ThrowingSupplier<Unit, Exception>>(), anyString())
    verify(mockDBOS, times(3)).setEvent(eq(STEPS_EVENT), any(Int::class.java))
  }
}
